### PR TITLE
Optimizes DonutProcessor token2json method for speed

### PR DESCRIPTION
**Pull request by michaelnation26**
# What does this PR do?

Speeds up the `token2json` method in `DonutProcessor` by calling `self.tokenizer.get_added_vocab()` once and reusing the `added_vocab` results in subsequent recursive calls. 

Fixes # (issue)

[Issue #20238](https://github.com/huggingface/transformers/issues/20238)

The `self.tokenizer.get_added_vocab()` call is somewhat expensive. It takes 50 - 70ms to complete. In the initial implementation of the `token2json` method, `self.tokenizer.get_added_vocab()` is called for every XML tag. If there are 10 XML tags, it would take 500 - 700ms in total. 

This PR makes the `token2json` method run time constant at 50 - 70ms, regardless of the number of XML tags.

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [x] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
[Issue #20238](https://github.com/huggingface/transformers/issues/20238)
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [x] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

@NielsRogge

### CODE REVIEW

File: tests/models/donut/test_processing_donut.py [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+
+import unittest
+
+from transformers import DonutProcessor
+
+
+DONUT_PRETRAINED_MODEL_NAME = "nielsr/donut-base"
+
+
+class DonutProcessorTest(unittest.TestCase):
```
Comment (michaelnation26) [inline tests/models/donut/test_processing_donut.py:25]: Do I need to add the `@require_torch` and `@require_vision` decorators here?

File: tests/models/donut/test_processing_donut.py [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+
+import unittest
+
+from transformers import DonutProcessor
```
Comment (michaelnation26) [inline tests/models/donut/test_processing_donut.py:19]: Should the `is_vision_available` method be checked first?
```
if is_vision_available():
    from transformers import DonutProcessor
```

Comment (NielsRogge) [reply tests/models/donut/test_processing_donut.py:19]: No, the processor itself doesn't rely on a framework, only `DonutFeatureExtractor` (soon to be called `DonutImageProcessor` - we're gonna deprecate the term "feature extractor") does, as it relies on Pillow.

**Comment by michaelnation26** [comment]
Hi @NielsRogge . Do I need to make any changes to this PR? If not, can I please get an approval so we can merge the PR.

File: tests/models/donut/test_processing_donut.py [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+
+
+DONUT_PRETRAINED_MODEL_NAME = "nielsr/donut-base"
```
Comment (NielsRogge) [inline tests/models/donut/test_processing_donut.py:22]: Could you please update this to the official checkpoint: naver-clova-ix/donut-base

**Review by NielsRogge — APPROVED** [review]
Thanks for adding!

File: tests/models/donut/test_processing_donut.py (hunk 1 above)
Comment (sgugger) [reply tests/models/donut/test_processing_donut.py:25]: It looks like tests are all passing without, so no need to add those.

**Review by sgugger — APPROVED** [review]
Thanks for your contribution!

**Comment by michaelnation26** [comment]
Hi @sgugger and @NielsRogge ,  after updating the `DONUT_PRETRAINED_MODEL_NAME` value and committing, the `check_repository_consistency` was failing on CircleCI so I decided to run `git fetch upstream` and `git rebase upstream/main` by following these [instructions](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#create-a-pull-request). I may have botched the next steps. The git message said my branch diverged. I did a `git pull` and `git config pull.rebase true`, then another `git pull`. Finally, a `git push`. Now, my PR includes everyone else's changes from the rebase. Do I leave these new changes in my PR or remove them? I only know the basics of Git so I'm not sure what to do next. Also, a couple of tests are now failing that are not related to my changes. Not sure how to resolve those.

**Comment by sgugger** [comment]
Can you fix the last conflict and force-push this time? It might work and remove the huge diff. Otherwise you'll need to close this PR and open a fresh one.

**Review by sgugger — APPROVED** [review]
Perfect, thanks!

**Pull request merged**